context_header.go

```diff
@@ -36,31 +36,56 @@ type ContextWithHeaders interface {
 	SetResponseHeaders(map[string]string)
 }
 
+type headersContextKey int
+
+const headersKey headersContextKey = iota
+
 type headerCtx struct {
 	context.Context
+}
+
+// headersContainer stores the headers, and is itself stored in the context under `headersKey`
+type headersContainer struct {
 	reqHeaders  map[string]string
 	respHeaders map[string]string
 }
 
+func (c *headerCtx) headers() *headersContainer {
+	if h, ok := c.Value(headersKey).(*headersContainer); ok {
```

Comment:
since the default value is `nil` if it's not found, I'm ok with simplifying this to
```go
return c.Value(headersKey).(*headersContainer)
```